ray-operator/controllers/ray/common/service_test.go

```diff
@@ -405,3 +402,136 @@ func TestBuildServiceForHeadPodPortsOrder(t *testing.T) {
+	expectedNamespace := serviceInstance.Namespace
+	if !reflect.DeepEqual(expectedNamespace, actualNamespace) {
+		t.Fatalf("Expected `%v` but got `%v`", expectedNamespace, actualNamespace)
+	}
+}
+
+func TestUserSpecifiedServeService(t *testing.T) {
+	// Use any RayService instance as a base for the test.
+	testRayServiceWithServeService := serviceInstance.DeepCopy()
+
+	userName := "user-custom-name"
+	userNamespace := "user-custom-namespace"
+	userLabels := map[string]string{"userLabelKey": "userLabelValue", RayClusterLabelKey: "userClusterName"} // Override default cluster name
+	userAnnotations := map[string]string{"userAnnotationKey": "userAnnotationValue", "userAnnotationKey2": "userAnnotationValue2"}
+	userPort := corev1.ServicePort{Name: "serve", Port: 12345}
+	userPortOverride := corev1.ServicePort{Name: DefaultClientPortName, Port: 98765} // Override default client port (10001)
+	userPorts := []corev1.ServicePort{userPort, userPortOverride}
+	userSelector := map[string]string{"userSelectorKey": "userSelectorValue", RayClusterLabelKey: "userSelectorClusterName"}
+	// Specify a "LoadBalancer" type, which differs from the default "ClusterIP" type.
+	userType := corev1.ServiceTypeLoadBalancer
+
+	testRayServiceWithServeService.Spec.ServeService = &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        userName,
+			Namespace:   userNamespace,
+			Labels:      userLabels,
+			Annotations: userAnnotations,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports:    userPorts,
+			Selector: userSelector,
+			Type:     userType,
+		},
+	}
+
+	svc, err := BuildServeServiceForRayService(*testRayServiceWithServeService, *instanceWithWrongSvc)
+	if err != nil {
+		t.Errorf("failed to build serve service: %v", err)
+	}
+
+	// Check every annotation is in the service annotation
+	for k := range userAnnotations {
+		if _, ok := svc.ObjectMeta.Annotations[k]; !ok {
+			t.Errorf("Final labels should contain key=%s", k)
+		}
+	}
+
+	// Check that selectors only have default selectors
+	if len(svc.Spec.Selector) != 2 {
+		t.Errorf("Selectors should have just 2 keys %s and %s", RayClusterLabelKey, RayClusterServingServiceLabelKey)
+	}
+	if svc.Spec.Selector[RayClusterLabelKey] != instanceWithWrongSvc.Name {
+		t.Errorf("Serve Service selector key %s value didn't match expected value : expected value=%s, actual value=%s", RayClusterLabelKey, instanceWithWrongSvc.Name, svc.Spec.Selector[RayClusterLabelKey])
+	}
+	if svc.Spec.Selector[RayClusterServingServiceLabelKey] != EnableRayClusterServingServiceTrue {
+		t.Errorf("Serve Service selector key %s value didn't match expected value : expected value=%s, actual value=%s", RayClusterServingServiceLabelKey, EnableRayClusterServingServiceTrue, svc.Spec.Selector[RayClusterServingServiceLabelKey])
+	}
+
+	// ports should only have DefaultServePort
+	ports := svc.Spec.Ports
+	expectedPortName := DefaultServingPortName
+	for _, port := range ports {
+		if port.Name != DefaultServingPortName {
```

Comment:
Should we also check the value of the port?

Reply:
Added a check for value of the port